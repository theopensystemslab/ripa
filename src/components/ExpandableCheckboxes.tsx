import { Button } from "@material-ui/core";
import Box from "@material-ui/core/Box";
import ExpansionPanel from "@material-ui/core/ExpansionPanel";
import ExpansionPanelDetails from "@material-ui/core/ExpansionPanelDetails";
import ExpansionPanelSummary from "@material-ui/core/ExpansionPanelSummary";
import FormControl from "@material-ui/core/FormControl";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import FormGroup from "@material-ui/core/FormGroup";
import FormLabel from "@material-ui/core/FormLabel";
import { makeStyles } from "@material-ui/core/styles";
import Typography from "@material-ui/core/Typography";
import { ExpandMore } from "@material-ui/icons";
import { useFormik } from "formik";
import React, { useState } from "react";

import Messages from "../shared/components/submit-messages";
import Checkbox from "./Checkbox";

const useStyles = makeStyles(theme => ({
  formControlLabelRoot: {
    marginLeft: 0,
    marginBottom: theme.spacing(1.5)
  },
  formControlLabel: {
    fontWeight: 400
  },
  expansionPanel: {
    background: "none"
  },
  expandedPanel: {
    backgroundColor: theme.palette.grey[100]
  },
  summary: {
    padding: theme.spacing(1, 2)
  },
  summaryContent: {
    margin: "0 !important"
  },
  details: {
    padding: theme.spacing(0, 2, 1)
  }
}));

interface IExpandableCheckboxes {
  title: string;
  name: string;
  panelsOptions: {
    sectionTitle: string;
    values: string[];
  }[];
  callback?;
  required?: boolean;
}

export const ExpandableCheckboxes: React.FC<IExpandableCheckboxes> = ({
  title = "",
  panelsOptions = [
    {
      sectionTitle: "",
      values: []
    }
  ],
  name = "",
  callback = undefined,
  required = false
}) => {
  const [successMessageVisible, setSuccessMessageVisible] = useState(false);
  const [submitButtonDisabled, setSubmitButtonDisabled] = useState(required);

  const formik = useFormik({
    initialValues: {
      [name]: "",
      noOfPanels: panelsOptions.length,
      selectedSections: [],
      selectedOptions: []
    },
    validate: values => {
      if (!required) return;

      const unDuplicatedSections = Array.from(new Set(values.selectedSections));
      if (unDuplicatedSections.length !== values.noOfPanels) {
        setSubmitButtonDisabled(true);
      } else {
        setSubmitButtonDisabled(false);
      }
    },
    onSubmit: (values, { resetForm }) => {
      setSuccessMessageVisible(true);
      setTimeout(() => {
        resetForm();
        setSuccessMessageVisible(false);
        setSubmitButtonDisabled(false);
      }, 1000);
    }
  });
  const classes = useStyles();
  return (
    <form
      data-testid="expandableForm"
      onSubmit={callback || formik.handleSubmit}
    >
      <Box maxWidth={500}>
        <Typography variant="h4" gutterBottom>
          <strong>{title}</strong>
        </Typography>
        {panelsOptions.map(({ sectionTitle, values }, index) => (
          <ExpansionPanel
            key={`${title}-${index}`}
            classes={{
              root: classes.expansionPanel,
              expanded: classes.expandedPanel
            }}
          >
            <ExpansionPanelSummary
              expandIcon={<ExpandMore />}
              aria-label="Expand"
              aria-controls="additional-actions1-content"
              id="additional-actions1-header"
              classes={{
                root: classes.summary,
                content: classes.summaryContent
              }}
            >
              {sectionTitle}
            </ExpansionPanelSummary>
            <ExpansionPanelDetails className={classes.details}>
              <FormControl component="fieldset">
                <FormLabel component="legend" />
                <FormGroup>
                  {values.map((checkbox, index) => (
                    <div key={`${checkbox}-${index}`}>
                      <FormControlLabel
                        classes={{
                          root: classes.formControlLabelRoot,
                          label: classes.formControlLabel
                        }}
                        control={
                          <Checkbox
                            checked={formik.values.selectedOptions.includes(
                              checkbox
                            )}
                            onChange={e => {
                              if (e.target.checked) {
                                formik.setFieldValue("selectedOptions", [
                                  ...formik.values.selectedOptions,
                                  checkbox
                                ]);
                                formik.setFieldValue("selectedSections", [
                                  ...formik.values.selectedSections,
                                  sectionTitle
                                ]);
                              } else {
                                formik.setFieldValue(
                                  "selectedOptions",
                                  formik.values.selectedOptions.filter(
                                    el => el !== checkbox
                                  )
                                );
                                formik.setFieldValue(
                                  "selectedSections",
                                  formik.values.selectedSections.filter(
                                    el => el !== sectionTitle
                                  )
                                );
                              }
                            }}
                            value={checkbox}
                            name={checkbox}
                          />
                        }
                        label={checkbox}
                      />
                    </div>
                  ))}
                </FormGroup>
              </FormControl>
            </ExpansionPanelDetails>
          </ExpansionPanel>
        ))}
        <Box pt={3}>
          <Button
            type="submit"
            variant="contained"
            color="primary"
            disabled={submitButtonDisabled}
            style={{ margin: "10px 0" }}
          >
            Save and Continue
          </Button>
        </Box>
        <div>
          {successMessageVisible ? (
            <Messages type="success" message="Form submitted successfully" />
          ) : null}
        </div>
      </Box>
    </form>
  );
};

export default ExpandableCheckboxes;
