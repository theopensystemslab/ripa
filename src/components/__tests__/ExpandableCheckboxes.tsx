import { cleanup, fireEvent, render } from "@testing-library/react";
import * as React from "react";

import ExpandableCheckboxes from "../ExpandableCheckboxes";

afterEach(cleanup);

describe("Expandable Checkboxes Component", () => {
  let title: string;
  let name: string;
  let panelsOptions: { sectionTitle: string; values: string[] }[];
  beforeEach(() => {
    title = "Expandable Checkboxes";
    name = "ExpandableCheckboxes";
    panelsOptions = [
      {
        sectionTitle: "section A",
        values: ["Response A1", "Response A2", "Response A3"]
      },
      {
        sectionTitle: "section B",
        values: ["Response B1", "Response B2", "Response B3"]
      },
      {
        sectionTitle: "section C",
        values: ["Response C1", "Response C2", "Response C3"]
      }
    ];
  });
  it("should render success message when submitted ", async () => {
    const { getByLabelText, findByText, getByTestId } = render(
      <ExpandableCheckboxes
        title={title}
        name={name}
        panelsOptions={panelsOptions}
      />
    );
    fireEvent.click(getByLabelText("Response A1"));
    fireEvent.submit(getByTestId("expandableForm"));
    expect(await findByText(/form submitted successfully/i)).toBeInTheDocument;
  });
  it("should render snapshot", () => {
    const { asFragment } = render(
      <ExpandableCheckboxes
        title={title}
        name={name}
        panelsOptions={panelsOptions}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
