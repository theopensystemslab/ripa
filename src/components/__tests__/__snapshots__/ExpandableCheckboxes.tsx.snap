// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expandable Checkboxes Component should render snapshot 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-372"
    tabindex="0"
  >
    <form
      data-testid="expandableForm"
    >
      <div
        class="MuiBox-root MuiBox-root-374"
      >
        <div
          class="MuiBox-root MuiBox-root-375"
        >
          <div
            class="makeStyles-root-376"
          >
            <div
              class="MuiTypography-root makeStyles-question-377 MuiTypography-h4"
            >
              <strong>
                Expandable Checkboxes
              </strong>
            </div>
          </div>
        </div>
        <div
          class="MuiPaper-root MuiExpansionPanel-root makeStyles-expansionPanel-367 MuiExpansionPanel-rounded MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            aria-controls="additional-actions1-content"
            aria-disabled="false"
            aria-expanded="false"
            aria-label="Expand"
            class="MuiButtonBase-root MuiExpansionPanelSummary-root makeStyles-summary-369"
            id="additional-actions1-header"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiExpansionPanelSummary-content makeStyles-summaryContent-370"
            >
              section A
            </div>
            <div
              aria-disabled="false"
              aria-hidden="true"
              class="MuiButtonBase-root MuiIconButton-root MuiExpansionPanelSummary-expandIcon MuiIconButton-edgeEnd"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
          </div>
          <div
            class="MuiCollapse-container MuiCollapse-hidden"
            style="min-height: 0px;"
          >
            <div
              class="MuiCollapse-wrapper"
            >
              <div
                class="MuiCollapse-wrapperInner"
              >
                <div
                  aria-labelledby="additional-actions1-header"
                  id="additional-actions1-content"
                  role="region"
                >
                  <div
                    class="MuiExpansionPanelDetails-root makeStyles-details-371"
                  >
                    <fieldset
                      class="MuiFormControl-root"
                    >
                      <legend
                        class="MuiFormLabel-root"
                      />
                      <div
                        class="MuiFormGroup-root"
                      >
                        <div>
                          <label
                            class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-365"
                          >
                            <span
                              aria-disabled="false"
                              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-503 MuiCheckbox-root makeStyles-checkBoxRoot-494 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                            >
                              <span
                                class="MuiIconButton-label"
                              >
                                <input
                                  class="PrivateSwitchBase-input-506"
                                  data-indeterminate="false"
                                  name="Response A1"
                                  type="checkbox"
                                  value="Response A1"
                                />
                                <span
                                  class="makeStyles-icon-495"
                                  font-size="default"
                                />
                              </span>
                              <span
                                class="MuiTouchRipple-root"
                              />
                            </span>
                            <span
                              class="MuiTypography-root MuiFormControlLabel-label makeStyles-formControlLabel-366 MuiTypography-body1"
                            >
                              Response A1
                            </span>
                          </label>
                        </div>
                        <div>
                          <label
                            class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-365"
                          >
                            <span
                              aria-disabled="false"
                              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-503 MuiCheckbox-root makeStyles-checkBoxRoot-494 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                            >
                              <span
                                class="MuiIconButton-label"
                              >
                                <input
                                  class="PrivateSwitchBase-input-506"
                                  data-indeterminate="false"
                                  name="Response A2"
                                  type="checkbox"
                                  value="Response A2"
                                />
                                <span
                                  class="makeStyles-icon-495"
                                  font-size="default"
                                />
                              </span>
                              <span
                                class="MuiTouchRipple-root"
                              />
                            </span>
                            <span
                              class="MuiTypography-root MuiFormControlLabel-label makeStyles-formControlLabel-366 MuiTypography-body1"
                            >
                              Response A2
                            </span>
                          </label>
                        </div>
                        <div>
                          <label
                            class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-365"
                          >
                            <span
                              aria-disabled="false"
                              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-503 MuiCheckbox-root makeStyles-checkBoxRoot-494 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                            >
                              <span
                                class="MuiIconButton-label"
                              >
                                <input
                                  class="PrivateSwitchBase-input-506"
                                  data-indeterminate="false"
                                  name="Response A3"
                                  type="checkbox"
                                  value="Response A3"
                                />
                                <span
                                  class="makeStyles-icon-495"
                                  font-size="default"
                                />
                              </span>
                              <span
                                class="MuiTouchRipple-root"
                              />
                            </span>
                            <span
                              class="MuiTypography-root MuiFormControlLabel-label makeStyles-formControlLabel-366 MuiTypography-body1"
                            >
                              Response A3
                            </span>
                          </label>
                        </div>
                      </div>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiPaper-root MuiExpansionPanel-root makeStyles-expansionPanel-367 MuiExpansionPanel-rounded MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            aria-controls="additional-actions1-content"
            aria-disabled="false"
            aria-expanded="false"
            aria-label="Expand"
            class="MuiButtonBase-root MuiExpansionPanelSummary-root makeStyles-summary-369"
            id="additional-actions1-header"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiExpansionPanelSummary-content makeStyles-summaryContent-370"
            >
              section B
            </div>
            <div
              aria-disabled="false"
              aria-hidden="true"
              class="MuiButtonBase-root MuiIconButton-root MuiExpansionPanelSummary-expandIcon MuiIconButton-edgeEnd"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
          </div>
          <div
            class="MuiCollapse-container MuiCollapse-hidden"
            style="min-height: 0px;"
          >
            <div
              class="MuiCollapse-wrapper"
            >
              <div
                class="MuiCollapse-wrapperInner"
              >
                <div
                  aria-labelledby="additional-actions1-header"
                  id="additional-actions1-content"
                  role="region"
                >
                  <div
                    class="MuiExpansionPanelDetails-root makeStyles-details-371"
                  >
                    <fieldset
                      class="MuiFormControl-root"
                    >
                      <legend
                        class="MuiFormLabel-root"
                      />
                      <div
                        class="MuiFormGroup-root"
                      >
                        <div>
                          <label
                            class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-365"
                          >
                            <span
                              aria-disabled="false"
                              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-503 MuiCheckbox-root makeStyles-checkBoxRoot-494 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                            >
                              <span
                                class="MuiIconButton-label"
                              >
                                <input
                                  class="PrivateSwitchBase-input-506"
                                  data-indeterminate="false"
                                  name="Response B1"
                                  type="checkbox"
                                  value="Response B1"
                                />
                                <span
                                  class="makeStyles-icon-495"
                                  font-size="default"
                                />
                              </span>
                              <span
                                class="MuiTouchRipple-root"
                              />
                            </span>
                            <span
                              class="MuiTypography-root MuiFormControlLabel-label makeStyles-formControlLabel-366 MuiTypography-body1"
                            >
                              Response B1
                            </span>
                          </label>
                        </div>
                        <div>
                          <label
                            class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-365"
                          >
                            <span
                              aria-disabled="false"
                              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-503 MuiCheckbox-root makeStyles-checkBoxRoot-494 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                            >
                              <span
                                class="MuiIconButton-label"
                              >
                                <input
                                  class="PrivateSwitchBase-input-506"
                                  data-indeterminate="false"
                                  name="Response B2"
                                  type="checkbox"
                                  value="Response B2"
                                />
                                <span
                                  class="makeStyles-icon-495"
                                  font-size="default"
                                />
                              </span>
                              <span
                                class="MuiTouchRipple-root"
                              />
                            </span>
                            <span
                              class="MuiTypography-root MuiFormControlLabel-label makeStyles-formControlLabel-366 MuiTypography-body1"
                            >
                              Response B2
                            </span>
                          </label>
                        </div>
                        <div>
                          <label
                            class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-365"
                          >
                            <span
                              aria-disabled="false"
                              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-503 MuiCheckbox-root makeStyles-checkBoxRoot-494 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                            >
                              <span
                                class="MuiIconButton-label"
                              >
                                <input
                                  class="PrivateSwitchBase-input-506"
                                  data-indeterminate="false"
                                  name="Response B3"
                                  type="checkbox"
                                  value="Response B3"
                                />
                                <span
                                  class="makeStyles-icon-495"
                                  font-size="default"
                                />
                              </span>
                              <span
                                class="MuiTouchRipple-root"
                              />
                            </span>
                            <span
                              class="MuiTypography-root MuiFormControlLabel-label makeStyles-formControlLabel-366 MuiTypography-body1"
                            >
                              Response B3
                            </span>
                          </label>
                        </div>
                      </div>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiPaper-root MuiExpansionPanel-root makeStyles-expansionPanel-367 MuiExpansionPanel-rounded MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            aria-controls="additional-actions1-content"
            aria-disabled="false"
            aria-expanded="false"
            aria-label="Expand"
            class="MuiButtonBase-root MuiExpansionPanelSummary-root makeStyles-summary-369"
            id="additional-actions1-header"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiExpansionPanelSummary-content makeStyles-summaryContent-370"
            >
              section C
            </div>
            <div
              aria-disabled="false"
              aria-hidden="true"
              class="MuiButtonBase-root MuiIconButton-root MuiExpansionPanelSummary-expandIcon MuiIconButton-edgeEnd"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
          </div>
          <div
            class="MuiCollapse-container MuiCollapse-hidden"
            style="min-height: 0px;"
          >
            <div
              class="MuiCollapse-wrapper"
            >
              <div
                class="MuiCollapse-wrapperInner"
              >
                <div
                  aria-labelledby="additional-actions1-header"
                  id="additional-actions1-content"
                  role="region"
                >
                  <div
                    class="MuiExpansionPanelDetails-root makeStyles-details-371"
                  >
                    <fieldset
                      class="MuiFormControl-root"
                    >
                      <legend
                        class="MuiFormLabel-root"
                      />
                      <div
                        class="MuiFormGroup-root"
                      >
                        <div>
                          <label
                            class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-365"
                          >
                            <span
                              aria-disabled="false"
                              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-503 MuiCheckbox-root makeStyles-checkBoxRoot-494 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                            >
                              <span
                                class="MuiIconButton-label"
                              >
                                <input
                                  class="PrivateSwitchBase-input-506"
                                  data-indeterminate="false"
                                  name="Response C1"
                                  type="checkbox"
                                  value="Response C1"
                                />
                                <span
                                  class="makeStyles-icon-495"
                                  font-size="default"
                                />
                              </span>
                              <span
                                class="MuiTouchRipple-root"
                              />
                            </span>
                            <span
                              class="MuiTypography-root MuiFormControlLabel-label makeStyles-formControlLabel-366 MuiTypography-body1"
                            >
                              Response C1
                            </span>
                          </label>
                        </div>
                        <div>
                          <label
                            class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-365"
                          >
                            <span
                              aria-disabled="false"
                              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-503 MuiCheckbox-root makeStyles-checkBoxRoot-494 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                            >
                              <span
                                class="MuiIconButton-label"
                              >
                                <input
                                  class="PrivateSwitchBase-input-506"
                                  data-indeterminate="false"
                                  name="Response C2"
                                  type="checkbox"
                                  value="Response C2"
                                />
                                <span
                                  class="makeStyles-icon-495"
                                  font-size="default"
                                />
                              </span>
                              <span
                                class="MuiTouchRipple-root"
                              />
                            </span>
                            <span
                              class="MuiTypography-root MuiFormControlLabel-label makeStyles-formControlLabel-366 MuiTypography-body1"
                            >
                              Response C2
                            </span>
                          </label>
                        </div>
                        <div>
                          <label
                            class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-365"
                          >
                            <span
                              aria-disabled="false"
                              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-503 MuiCheckbox-root makeStyles-checkBoxRoot-494 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                            >
                              <span
                                class="MuiIconButton-label"
                              >
                                <input
                                  class="PrivateSwitchBase-input-506"
                                  data-indeterminate="false"
                                  name="Response C3"
                                  type="checkbox"
                                  value="Response C3"
                                />
                                <span
                                  class="makeStyles-icon-495"
                                  font-size="default"
                                />
                              </span>
                              <span
                                class="MuiTouchRipple-root"
                              />
                            </span>
                            <span
                              class="MuiTypography-root MuiFormControlLabel-label makeStyles-formControlLabel-366 MuiTypography-body1"
                            >
                              Response C3
                            </span>
                          </label>
                        </div>
                      </div>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-507"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
            style="margin: 10px 0px;"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Save and Continue
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div />
      </div>
    </form>
  </div>
</DocumentFragment>
`;
